Move and re-use persist_extra_files

mvdbeek opened the pull request:
This is in preparation of improving the directory datatype.

#BIOHACKEU23

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Move and re-use persist_extra_files

nsoranzo commented on lib/galaxy/job_execution/output_collect.py:
```diff
@@ -56,6 +60,9 @@
 )
 
+if TYPE_CHECKING:
+    from galaxy.model import DatasetInstance
```
We import a bunch of other classes from `galaxy.model` above, so probably we can import `DatasetInstance` there as well, if you like.

mvdbeek replied:
Moved it back to the top

nsoranzo commented on lib/galaxy/objectstore/__init__.py:
```diff
@@ -1577,3 +1585,27 @@ def set_dataset_object_store_id(self, dataset, require_shareable=True):
+
+
+def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance"):
+    if src_extra_files_path and os.path.exists(src_extra_files_path):
```
I don't think `src_extra_files_path` can be an empty string.

```suggestion
    if os.path.exists(src_extra_files_path):
```

mvdbeek replied:
It certainly could be in the past if the extra files dir hadn't been created yet. Not sure I'm ready remove the safe-guard ?

mvdbeek committed:
Fix wrong relpath calculation if object store cache is used

mvdbeek committed:
Fix typo-ish bug with persisting collection element extra files

I suppose we have no way to express in tools that we are discovering
collection elements with extra files, and that's why that hasn't caused
any problems ? Caught by 🦅👀!

nsoranzo commented on lib/galaxy/objectstore/__init__.py:
```diff
@@ -1577,3 +1585,25 @@ def set_dataset_object_store_id(self, dataset, require_shareable=True):
         self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
+
+
+def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance"):
```
```suggestion
def persist_extra_files(object_store: ObjectStore, src_extra_files_path: str, primary_data: "DatasetInstance") -> None:
```

nsoranzo commented on lib/galaxy/job_execution/output_collect.py:
```diff
@@ -723,9 +727,11 @@ def default_exit_code_file(files_dir, id_tag):
 
 
-def collect_extra_files(object_store, dataset, job_working_directory):
+def collect_extra_files(object_store: ObjectStore, dataset: "DatasetInstance", job_working_directory: str):
```
```suggestion
def collect_extra_files(object_store: ObjectStore, dataset: "DatasetInstance", job_working_directory: str) -> None:
```

mvdbeek committed:
Add return type annotation for collect_extra_files

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>
